rfcs/2022-11-04-automatic-component-validation.md

```diff
@@ -0,0 +1,512 @@
+the design, as it stands, having no clear way to automatically derive the implementation of the
+proposed trait from existing data, forcing all implementations to be written manually.
+
+## Prior Art
+
+Vector already has existing mechanisms for validating components in terms of the Component
+Specification, and so on. This comes in the form of familiar helper methods like
+`assert_sink_compliance`, `assert_source_error`, etc. Those helpers operate with a similar approach
+to this proposal in that we run the relevant code in a way that ensures we can observe the state of
+the system before and after the code runs, calculating the validation results at the end.
+
+We do not have an existing mechanism for asserting the _correctness_ of internal telemetry, only
+that specific bits of internal telemetry are being emitted as intended. This is not necessarily a
+technical limitation, but a practical limitation of not yet having gotten to the point of adding
+support to validate the correctness.
+
+What both of these things do not provide, that this proposal does, is a mechanism to _ensure_ that a
+component that can be used has been actually tested against the intended criteria/specifications/etc
+and met them. While this doesn't happen as a result of the compiler/type system, and so is still
+somewhat fallible at the edges -- we have to actually do work to integrate this proposal with CI,
+etc -- it achieves uniform coverage across all components and provides the scaffolding to avoid the
+human fallibility aspect of "oops, we're not actually validating this component".
+
+It also heavily centralizes this validation logic such that any future work around updating what
+gets validated, or exposing more information during the validation step, and so on, can happen as a
+set of changes to a relatively small set of files, rather than, again, a change to every component
+involved, which is exactly where human fallibility has struck us in the past.
+
+## Alternatives
+
+One alternative approach could be to actually build Vector and run an entire configuration, much
+like an integration test, and we would send known data to Vector, collect the output, so on and so
+forth, much like this proposal... but we would simply collect it externally, much like `lading` does
+in our soak tests.
+
+The biggest reason not to go with this alternative approach is that it still involves much of the
+scaffolding present in this proposal, such as getting a list of all components that should be
+tested, and writing code that can emit that "external resource" information, since a Vector
+configuration may depend heavily on default values, and so we would not always be able to extract
+all of the necessary information from a Vector configuration that was otherwise valid by itself.
+
+We'd simply be writing all of the same code, but bundling it into a separate binary that ran Vector,
+without -- at least as far as I can see -- a meaningful benefit to the isolation.
```

Comment:
Are there downsides for including this additional code in Vector in terms of binary size/compile time/etc - or should things be reasonably gated by features?

Reply:
This code, if added to a binary where all components are being compiled in, at least as-is, would not bring in any additional dependencies, so the difference should be imperceptible compared to a normal full build i.e. `cargo build`.